File: vaultlib_go/console_tests/promise.go
//go:build js && wasm

package consoletests

import (
	"errors"
	"syscall/js"

	"wraith.me/vaultlib/jsutil"
)

// promiseVanilla(x: number, y: number): Promise<number>
func TestPromiseVanilla(this js.Value, args []js.Value) interface{} {
	//Get the input args
	x := int(args[0].Float())
	y := int(args[1].Float())

	promiseConstructor := js.Global().Get("Promise")
	return promiseConstructor.New(js.FuncOf(func(this js.Value, args []js.Value) interface{} {
		//Setup the resolve and reject functions
		resolve := args[0]
		reject := args[1]

		//Define the async function to run
		go func() {
			//Error condition
			if x < 0 || y < 0 {
				errorMsg := js.ValueOf("Input numbers must be non-negative")
				reject.Invoke(errorMsg)
				return
			}

			//Ok condition
			result := x * y
			resolve.Invoke(js.ValueOf(result))
		}()

		//Don't return anything
		return nil
	}))
}

// promise(x: number, y: number): Promise<number>
func TestPromise(this js.Value, args []js.Value) interface{} {
	//Get the input args
	x := int(args[0].Float())
	y := int(args[1].Float())

	//Setup the promise action to run
	fun := func() (js.Value, error) {
		if x < 0 || y < 0 {
			return jsutil.Nil, errors.New("Input numbers must be non-negative")
		}

		//Ok condition
		return js.ValueOf(x * y), nil
	}

	//Construct and return the promise
	return jsutil.Promise(fun)
}

File: vaultlib_go/console_tests/promise.js
// Test case 1: Positive numbers
promise(5, 4)
	.then(result => console.log("Test 1 (5 * 4) Result:", result))
	.catch(err => console.error("Test 1 Error:", err));

// Test case 2: Zero
promise(0, 10)
	.then(result => console.log("Test 2 (0 * 10) Result:", result))
	.catch(err => console.error("Test 2 Error:", err));

// Test case 3: Negative number (should reject)
promise(-3, 7)
	.then(result => console.log("Test 3 (-3 * 7) Result:", result))
	.catch(err => console.error("Test 3 Error:", err));

// Test case 4: Large numbers
promise(1000, 2000)
	.then(result => console.log("Test 4 (1000 * 2000) Result:", result))
	.catch(err => console.error("Test 4 Error:", err));

// Test case 5: Both negative numbers (should reject)
promise(-5, -5)
	.then(result => console.log("Test 5 (-5 * -5) Result:", result))
	.catch(err => console.error("Test 5 Error:", err));

// Test case 6: Non-integer input (behavior depends on your implementation)
promise(3.14, 2)
	.then(result => console.log("Test 6 (3.14 * 2) Result:", result))
	.catch(err => console.error("Test 6 Error:", err));

console.log("All test cases have been initiated. Check the results above.");
